colcon_core/package_descriptor.py

```diff
@@ -144,11 +138,19 @@ def get_recursive_dependencies(
         return dependencies
 
     def __hash__(self):  # noqa: D105
-        return hash((self.realpath, self.type, self.name))
+        return hash((self.type, self.name))
 
     def __eq__(self, other):  # noqa: D105
-        return (self.realpath, self.type, self.name) == \
-            (other.realpath, other.type, other.name)
+        if type(self) != type(other):
+            return NotImplemented
+        if self is other:
+            return True
+        if not (self.type == other.type and self.name == other.name):
+            return False
+        if self.path == other.path:
+            return True
+
+        return self.realpath == other.realpath
```

Comment:
Some of these conditions can probably be consolidated put lets focus on the main part where `realpath()` is evaluated first.

Reply:
I have it re-evaluated when the property is accessed because the courser optimizations make repeated calls rare enough that it's not worth optimizing. You need only call `realpath` in the case where the source tree has two different paths to the same project, as opposed to previously where it was needed at least twice (once for finding the hash index, at least once for probing equality) for every operation involving finding a `PackageDescriptor` in a hash table.

Reply:
Please insert the following comment above this line: `# check realpath last since it is the most expensive to compute`.